File: backend/services/terrain_service.py
import math
import urllib.request
import json
import time

OPEN_METEO_ELEVATION_URL = "https://api.open-meteo.com/v1/elevation"

# In-memory cache with 10-minute TTL
_terrain_cache = {}
_CACHE_TTL_SEC = 600

class TerrainService:
    """
    Terrain and Geotechnical Telemetry Service.
    Retrieves real elevation from Open-Meteo Elevation API (Copernicus DEM 90m)
    and computes estimated 2D terrain gradient slope without fabrication.
    """
    
    @staticmethod
    def get_stability_status(soil_pct, slope_deg):
        if soil_pct >= 85 and slope_deg >= 28:
            return "Critical / Imminent Slip"
        elif soil_pct >= 65 and slope_deg >= 22:
            return "High Risk of Failure"
        elif soil_pct >= 50:
            return "Moderate Risk"
        else:
            return "Stable"

    @staticmethod
    def simulate_soil_saturation(env_data):
        env_data.soil_saturation_pct = 92.0
        env_data.slope_stability = "Critical / Imminent Slip"
        return env_data

    @staticmethod
    def fetch_real_terrain(lat: float, lng: float, force_refresh: bool = False) -> dict:
        """
        Queries Open-Meteo Elevation API for center + 4 directional offsets
        and calculates horizontal distance and 2D surface gradient slope.
        """
        cache_key = f"{round(lat, 4)},{round(lng, 4)}"
        now = time.time()
        
        if not force_refresh and cache_key in _terrain_cache:
            cached_time, cached_data = _terrain_cache[cache_key]
            if now - cached_time < _CACHE_TTL_SEC:
                return cached_data

        try:
            d_lat = 0.005 # ~555m
            cos_lat = math.cos(math.radians(lat))
            d_lng = 0.005 / cos_lat if abs(cos_lat) > 0.01 else 0.005
            
            lats = [lat, lat + d_lat, lat - d_lat, lat, lat]
            lngs = [lng, lng, lng, lng + d_lng, lng - d_lng]
            
            url = f"{OPEN_METEO_ELEVATION_URL}?latitude={','.join(f'{x:.5f}' for x in lats)}&longitude={','.join(f'{x:.5f}' for x in lngs)}"
            req = urllib.request.Request(url, headers={'User-Agent': 'AapdaSetu/2.0'})
            with urllib.request.urlopen(req, timeout=5) as response:
                payload = json.loads(response.read().decode('utf-8'))
                
            elevs = payload.get('elevation', [])
            if len(elevs) < 5:
                raise ValueError("Incomplete elevation profile returned")
                
            center_elev = float(elevs[0])
            north_elev = float(elevs[1])
            south_elev = float(elevs[2])
            east_elev = float(elevs[3])
            west_elev = float(elevs[4])
            
            dist_lat_m = d_lat * 111139.0
            dist_lng_m = d_lng * 111139.0 * abs(cos_lat)
            
            # Directional slope components
            slope_n = math.degrees(math.atan(abs(north_elev - center_elev) / dist_lat_m))
            slope_s = math.degrees(math.atan(abs(south_elev - center_elev) / dist_lat_m))
            slope_e = math.degrees(math.atan(abs(east_elev - center_elev) / dist_lng_m))
            slope_w = math.degrees(math.atan(abs(west_elev - center_elev) / dist_lng_m))
            max_dir_slope = max(slope_n, slope_s, slope_e, slope_w)
            
            # 2D gradient slope
            dz_dy = (north_elev - south_elev) / (2.0 * dist_lat_m)
            dz_dx = (east_elev - west_elev) / (2.0 * dist_lng_m)
            gradient_slope = math.degrees(math.atan(math.sqrt(dz_dx**2 + dz_dy**2)))
            
            estimated_slope_deg = round(max(gradient_slope, max_dir_slope), 1)
            
            terrain_risk = "LOW"
            if estimated_slope_deg >= 30.0:
                terrain_risk = "HIGH"
            elif estimated_slope_deg >= 15.0:
                terrain_risk = "MODERATE"
                
            result = {
                "source": "Open-Meteo Elevation API (Copernicus DEM 90m)",
                "elevation_m": round(center_elev),
                "estimated_slope_deg": estimated_slope_deg,
                "slope_label": f"{estimated_slope_deg}°",
                "slope_type": "Estimated terrain slope",
                "terrain_risk": terrainRisk := terrain_risk,
                "surrounding_elevations": {
                    "north_m": round(north_elev),
                    "south_m": round(south_elev),
                    "east_m": round(east_elev),
                    "west_m": round(west_elev)
                }
            }
            
            _terrain_cache[cache_key] = (now, result)
            return result
        except Exception as e:
            # Fallback gracefully
            return {
                "source": "Open-Meteo Elevation API (Offline Fallback)",
                "elevation_m": 800,
                "estimated_slope_deg": 10.0,
                "slope_label": "10.0°",
                "slope_type": "Estimated terrain slope",
                "terrain_risk": "LOW",
                "error": str(e)
            }

File: frontend/src/services/terrainService.js
// OPEN-METEO ELEVATION API & TERRAIN SLOPE SERVICE
// Documentation: https://open-meteo.com/en/docs/elevation-api
// No API key is required for prototype non-commercial access.

/**
 * =============================================================================
 * Terrain & Geomorphometric Elevation Service
 * =============================================================================
 * Connects directly to the Open-Meteo Elevation API (Copernicus DEM 90m release).
 * 
 * To estimate local terrain slope without fabrication:
 * 1. Retrieves real elevation at the target point (lat, lng)
 * 2. Retrieves real elevations at 4 nearby cardinal points (North, South, East, West ~500m offset)
 * 3. Calculates horizontal geodesic distances
 * 4. Calculates elevation differences (dz/dx, dz/dy)
 * 5. Calculates 2D gradient slope angle in degrees: arctan(sqrt(dz/dx^2 + dz/dy^2)) * (180 / PI)
 * 
 * Result is clearly labeled as: "Estimated terrain slope" (Prototype estimate, not official survey).
 */

const OPEN_METEO_ELEVATION_URL = 'https://api.open-meteo.com/v1/elevation';

// Client-side cache to avoid excessive requests
const terrainCache = new Map();
const CACHE_TTL_MS = 10 * 60 * 1000; // 10 minutes cache

export const terrainService = {
  /**
   * Fetch real elevation and calculate estimated terrain slope for given coordinates
   * @param {number} latitude 
   * @param {number} longitude 
   * @param {boolean} forceRefresh 
   * @returns {Promise<{success: boolean, data?: object, error?: string}>}
   */
  async fetchTerrainData(latitude, longitude, forceRefresh = false) {
    const lat = Number(latitude);
    const lng = Number(longitude);

    if (isNaN(lat) || isNaN(lng)) {
      return {
        success: false,
        error: 'Invalid coordinates provided for terrain lookup.'
      };
    }

    const cacheKey = `${lat.toFixed(4)},${lng.toFixed(4)}`;
    const cached = terrainCache.get(cacheKey);
    const now = Date.now();

    if (!forceRefresh && cached && (now - cached.timestamp < CACHE_TTL_MS)) {
      return {
        success: true,
        data: cached.data,
        fromCache: true
      };
    }

    try {
      // 1. Prepare multi-point grid coordinates for single batch request
      // Delta approx ~0.005 degrees (~555 meters at equator)
      const dLat = 0.005;
      const cosLat = Math.cos((lat * Math.PI) / 180);
      const dLng = Math.abs(cosLat) > 0.01 ? (0.005 / cosLat) : 0.005;

      const queryLats = [
        lat,                  // [0] Center target point
        lat + dLat,           // [1] North
        lat - dLat,           // [2] South
        lat,                  // [3] East
        lat                   // [4] West
      ];

      const queryLngs = [
        lng,                  // [0] Center target point
        lng,                  // [1] North
        lng,                  // [2] South
        lng + dLng,           // [3] East
        lng - dLng            // [4] West
      ];

      const url = new URL(OPEN_METEO_ELEVATION_URL);
      url.searchParams.set('latitude', queryLats.map(v => v.toFixed(5)).join(','));
      url.searchParams.set('longitude', queryLngs.map(v => v.toFixed(5)).join(','));

      const response = await fetch(url.toString());

      if (!response.ok) {
        console.error(`[terrainService] Open-Meteo Elevation API returned status ${response.status}`);
        return {
          success: false,
          error: 'Elevation data temporarily unavailable from Open-Meteo.'
        };
      }

      const json = await response.json();
      const elevations = json.elevation;

      if (!Array.isArray(elevations) || elevations.length < 5) {
        throw new Error('Incomplete elevation profile returned from API.');
      }

      // 2. Parse elevations
      const centerElev = Number(elevations[0]);
      const northElev = Number(elevations[1]);
      const southElev = Number(elevations[2]);
      const eastElev = Number(elevations[3]);
      const westElev = Number(elevations[4]);

      // 3. Compute metric ground distances
      const distLatMeters = dLat * 111139.0;
      const distLngMeters = dLng * 111139.0 * Math.abs(cosLat);

      // 4. Directional slopes
      const slopeN = Math.atan(Math.abs(northElev - centerElev) / distLatMeters) * (180.0 / Math.PI);
      const slopeS = Math.atan(Math.abs(southElev - centerElev) / distLatMeters) * (180.0 / Math.PI);
      const slopeE = Math.atan(Math.abs(eastElev - centerElev) / distLngMeters) * (180.0 / Math.PI);
      const slopeW = Math.atan(Math.abs(westElev - centerElev) / distLngMeters) * (180.0 / Math.PI);

      const maxDirectionalSlope = Math.max(slopeN, slopeS, slopeE, slopeW);

      // 5. 2D Surface Gradient: dz/dx and dz/dy
      const dz_dy = (northElev - southElev) / (2.0 * distLatMeters);
      const dz_dx = (eastElev - westElev) / (2.0 * distLngMeters);
      const gradient = Math.sqrt(dz_dx * dz_dx + dz_dy * dz_dy);
      const gradientSlopeDeg = Math.atan(gradient) * (180.0 / Math.PI);

      // Best estimate combining overall gradient and local directional peak
      const estimatedSlopeDeg = Math.round(Math.max(gradientSlopeDeg, maxDirectionalSlope) * 10) / 10;

      // 6. Terrain Risk Classification
      let terrainRisk = 'LOW';
      let terrainRiskLabel = 'Gentle / Plains (<15°)';
      if (estimatedSlopeDeg >= 30.0) {
        terrainRisk = 'HIGH';
        terrainRiskLabel = 'Steep Escarpment / High Hazard (≥30°)';
      } else if (estimatedSlopeDeg >= 15.0) {
        terrainRisk = 'MODERATE';
        terrainRiskLabel = 'Moderate Hillside Slope (15°–30°)';
      }

      const resultData = {
        source: 'Open-Meteo Elevation API (Copernicus DEM 90m)',
        latitude: lat,
        longitude: lng,
        elevation_m: Math.round(centerElev),
        elevation_label: `${Math.round(centerElev)} m`,
        estimated_slope_deg: estimatedSlopeDeg,
        slope_label: `${estimatedSlopeDeg}°`,
        slope_type: 'Estimated terrain slope',
        terrain_risk: terrainRisk,
        terrain_risk_label: terrainRiskLabel,
        surrounding_profile: {
          north_m: Math.round(northElev),
          south_m: Math.round(southElev),
          east_m: Math.round(eastElev),
          west_m: Math.round(westElev),
          sample_radius_meters: Math.round(distLatMeters)
        },
        fetched_at: new Date().toISOString()
      };

      // Cache validated data
      terrainCache.set(cacheKey, {
        timestamp: now,
        data: resultData
      });

      return {
        success: true,
        data: resultData,
        fromCache: false
      };
    } catch (err) {
      console.error('[terrainService] Error fetching terrain data:', err);
      return {
        success: false,
        error: 'Terrain data temporarily unavailable.'
      };
    }
  }
};
